### PR TITLE
Move MPI-related code to C++

### PR DESCRIPTION
**Pull request by paulromano**
This pull request moves all MPI-related code from Fortran to C++. While most of it is straightforward, there are a few points worth mentioning:
- You'll notice that `global_tallies()` and `tally_results(int)` in the new `include/openmc/tallies/tally.h` file have strange return types. The more obvious thing to do (return an `xarray` or `xtensor`) actually doesn't work. The gist of it is that if you try to return an `xtensor` or an `xarray` that represents some other array that you have adapted (via `xt::adapt`), you end up getting a *copy*, so it's not possible to modify the array afterwards.
- I managed to get a `write_dataset` function that works for both `xarray`s and `xtensor`s, but it required making sure that the scalar version of `write_dataset` is enabled only for truly scalar values (hence the use of `std::enable_if_t<...>`).
- When converting `synchronize_bank`, I had to use `std::upper_bound` rather than `std::lower_bound` for binary searches to match the behavior of our existing binary search in the case where the value being searched is present in the array.

### CODE REVIEW

File: src/error.cpp [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+#include "openmc/message_passing.h"
+
+namespace openmc {
+
+#ifdef OPENMC_MPI
+void abort_mpi(int code)
+{
+    MPI_Abort(mpi::intracomm, code);
+}
+#endif
```
Comment (smharper) [inline src/error.cpp]: Is there any particular reason you put this implementation in a .cpp file rather than the .h?  Trying to avoid future circular dependencies with error.h and message_passing.h?  Otherwise, I think it might make sense to just leave this short implementation in the header file.

File: src/state_point.cpp [hunk 2]
```diff
@@ -1,16 +1,20 @@
...
 #include "openmc/settings.h"
+#include "openmc/simulation.h"
+#include "openmc/tallies/tally.h"
+
+#include "xtensor/xbuilder.hpp" // for empty_like
+#include "xtensor/xview.hpp"
+
+#include <algorithm>
+#include <string>
+#include <vector>
```
Comment (smharper) [inline src/state_point.cpp]: This `#include` order is reversed from our conventional standard library > third-party > openmc-specific ordering.  Do you want to move to this convention throughout the C++ code?  If so, it might be better to do that in a separate PR so we don't have two conventions coexisting.

File: include/openmc/tallies/tally.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+// Threadprivate variables
+
+extern "C" double global_tally_absorption;
+#pragma omp threadprivate(global_tally_absorption)
+
+//==============================================================================
+// Non-member functions
+//==============================================================================
+
+template <std::size_t N>
```
Comment (smharper) [inline include/openmc/tallies/tally.h]: Add a comment mentioning that `N` is the tensor dimension.

File: include/openmc/hdf5_interface.h [hunk 4]
```diff
@@ -333,7 +334,9 @@ write_attribute(hid_t obj_id, const char* name, const std::vector<T>& buffer)
 // Templates/overloads for write_dataset
 //==============================================================================
 
-template<typename T> inline void
+// Template for scalars (ensured by SFINAE)
+template<typename T> inline
+std::enable_if_t<std::is_scalar<std::decay_t<T>>::value>
 write_dataset(hid_t obj_id, const char* name, T buffer)
```
Comment (smharper) [inline include/openmc/hdf5_interface.h]: I find it confusing that the return type isn't explicit with this function.  Can you easily switch the `enable_if_t` to operate through a template parameter rather than the return type?

File: src/error.cpp (hunk 1 above)
Comment (paulromano) [reply src/error.cpp]: Yes, because this function is called from Fortran, it can't be inline. If the compiler were to optimizing away the function call, there would be no `abort_mpi` symbol for Fortran to call.

File: src/state_point.cpp (hunk 2 above)
Comment (paulromano) [reply src/state_point.cpp]: Yeah, that is my preferred convention now, but I can switch this back and handle all of them in a separate PR.

File: include/openmc/hdf5_interface.h (hunk 4 above)
Comment (paulromano) [reply include/openmc/hdf5_interface.h]: Yeah, there's a hidden default void in `std::enable_if_t`. I actually tried to do `enable_if_t` as template parameter, but for some reason that didn't work. This is my first time using this technique, so it's possible I did something wrong. If you know of a way to get it to work via a template parameter, I'm all ears.

**Comment by paulromano** [comment]
@smharper Regarding the name scoping issues -- things like `openmc_keff` were never really intended to be part of the C API. For that we have functions like `openmc_get_keff` which do have a `openmc_` prefix. Global variables like `keff` are namespaced in C++, so the only danger that arises is the fact that they are `extern "C"` (i.e., there's no mangling). That is a temporary issue though -- once we are fully C++, there is no need for `extern "C"` on anything other than C API functions. The benefit now is that when converting Fortran code, we don't need to add extra `openmc_` prefixes on everything because the normal scoping rules allow global variables in the `openmc::` namespace to be found.

File: include/openmc/hdf5_interface.h (hunk 4 above)
Comment (smharper) [reply include/openmc/hdf5_interface.h]: Hmm, I couldn't figure out the template parameter way either.  Oh well, it's only a small issue.

**Comment by smharper** [comment]
Checks are passing, code looks good, the PR's been open for over 36 hours.  Looks like everything is good for a merge!
